### PR TITLE
[mlr] implement native multicast routing for backbone router

### PR DESCRIPTION
**Pull request by simonlingoogle**
This commit implements native multicast routing using kernel Ip6 multicast forwarding. 

Backbone tests use latest commit of https://github.com/openthread/ot-br-posix/pull/604 to pass MLR checks.

### CODE REVIEW

File: src/posix/main.c
Comment (simonlingoogle) [inline src/posix/main.c:105]: Explain:
The change of this file is not mandatory for this PR. However, the original code is incorrect because `OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE` is never defined in main.c. This change allows us to test Backbone features using `ot-cli` program.

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:110]: why update `mListenerSet` regardless of `IsEnabled()`?

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:260]:
```suggestion
        // Forward multicast traffic from Backbone to Thread if the group address is subscribed by any Thread device via MLR.
```

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:268]: What's the plan for Outbound-forwarding flag and Admin-local forwarding flag defined in 9.4.7.1 Multicast Forwarding Configuration?

File: src/posix/platform/multicast_routing.hpp
Comment (librasungirl) [reply src/posix/platform/multicast_routing.hpp:173]: nit: is it necessary to use `Inbound` in the name?
It seems to me `bi-direction` forwarding here.

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:404]:
```suggestion
        MulticastRouteInfo   & routeInfo = it->second;
```

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:256]: It seems to me that now stale multicast route entries is only removed when there is new multicast traffic?
It might be more reasonable to flush stale entries periodically?

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:319]: what's the purpose of this `for` loop?
When adding new listener, it seems to me not possible that such entry exist already.

File: src/posix/main.c
Comment (simonlingoogle) [reply src/posix/main.c:105]: I guess `posix/main.c` is actually a application code thus can not get the core configuration?
Any comment on this? @bukepo

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:110]: I am assuming that it's possible to for SBBR to add multicast listener (at least in the future). So the current implementation does not assume that Multicast Listener is always added when it's Primary. But the `MulticastRoutingManager` is only enabled when it's Primary.

```
Spec 9.4.8.3.1:

Upon a successful registration of a Multicast Listener, the BBR sends out a BMLR.ntf
multicast on the Backbone Link interface to propagate all successfully processed multicast
registration entries to Secondary BBRs on the same Thread Network.
```

Note that the logic is same as before when using smcroute.

File: src/posix/platform/multicast_routing.hpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.hpp:168]: The current `Block` and `Remove` is actually accurate. The `Block` method does not add any MFC but unblock existing MFCs. Thus I tend to keep the current names.

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:462]: nit: move forward

https://github.com/openthread/openthread/blob/master/STYLE_GUIDE.md
> All scoped (i.e. stack) variable declarations should be placed together at the top of the enclosing scope in which they are used.
There shall be an empty line after all such variable declarations.

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:475]: typo?
```suggestion
            routeInfo.mValidPktCnt = validPktCnt;
```

File: src/posix/platform/multicast_routing.cpp
Comment (librasungirl) [inline src/posix/platform/multicast_routing.cpp:527]: missing `"%d"`?

File: src/posix/platform/multicast_routing.hpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.hpp:168]: I think `Unblock`, which is also used by smcroute, is better because people are usually unaware of `mrt->mfc_unres_queue` of the kernel.

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:211]: The current implementation is `{ return mMulticastRouterSock >= 0; }`, but it could change in the future. 

`FinalizeMulticastRouterSock` focus on the closing the socket if it's opened. I prefer `VerifyOrExit(mMulticastRouterSock != -1);` since it clearly states the condition of this method.

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:391]: Good catch!

The old implementation of this method is removing all MFCs. I failed to remove this line of code. :-1:

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:475]: Good catch!

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:110]: > It may require additional efforts for this optional feature far beyond update the listener set here comparatively.
Yes. But I expect SBBR will also add Multicast Listener just like PBBR does. The difference is that MulticastRoutingManager is not enabled for SBBR. So, whether or not `MulticastRoutingManager` is enabled, it tries to maintain `mListenerSet` so that it's exactly same as the Multicast Listeners Table.

> Besides, one concern is there might be some inconsistence between mListenerSet here and PBBR's multicast listener table if there is some primary/secondary state switch, might bring assert in Line 109.

Yes, this was my concern too. Actually the purpose of adding the address to `mListenerSet` even if `MulticastRoutingManager` is not enabled is to **make sure mListenerSet and Multicast Listeners Table are exactly same**.

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:258]: `expire per 60s` added

**Review by librasungirl — APPROVED** [review]
LGTM overall with one minor comments.

File: src/core/net/ip6_address.hpp
Comment (bukepo) [inline src/core/net/ip6_address.hpp:543]:
```suggestion
    void SetBytes(const uint8_t *aBytes) { memcpy(mFields.m8, aBytes, kSize); }
```

File: src/posix/platform/multicast_routing.cpp
Comment (bukepo) [inline src/posix/platform/multicast_routing.cpp]:
```suggestion
    otLogResultPlat(error, "MulticastRoutingManager: %s", __FUNCTION__);
exit:
    return;
```

File: src/posix/platform/multicast_routing.cpp
Comment (bukepo) [inline src/posix/platform/multicast_routing.cpp:83]: Bubble up the error?

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp]: revised

File: src/posix/platform/multicast_routing.cpp
Comment (bukepo) [inline src/posix/platform/multicast_routing.cpp:130]: avoid std::max since we intentionally removed dependency on stdc++ in #3130.

File: src/posix/platform/multicast_routing.cpp
Comment (bukepo) [inline src/posix/platform/multicast_routing.cpp:165]: VerifyOrDie? It seems to be a critical error.

File: src/posix/platform/multicast_routing.hpp
Comment (bukepo) [inline src/posix/platform/multicast_routing.hpp:112]: Is it possible to reuse otNetifIdentifier?

File: src/posix/platform/multicast_routing.hpp
Comment (bukepo) [inline src/posix/platform/multicast_routing.hpp:135]: avoid use stdc++?

File: src/posix/platform/platform-posix.h
Comment (simonlingoogle) [reply src/posix/platform/platform-posix.h]: I think `WARN` usually means the functionality is still working but with some restrictions. 
The log level should be `ERROR` because if it failed, the multicast routing is not working. 
But we don't have `ERROR`.

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp]: outdated

File: src/posix/platform/multicast_routing.hpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.hpp:112]: It does not make absolute sense to use otNetifIdentifier. The MifIndex is a index to the MIF in the kernel. We used 0 for Thread and 1 for Backbone. These values are different from `otNetifIdentifier` values. 
Since they are different values, I don't feel the necessity to reuse otNetifIdentifier.

BTW, these values are private to `MulticastRoutingManager`.

File: src/posix/platform/multicast_routing.hpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.hpp:135]: I don't have the strong intention to use stdc++. But I also not sure why using stdc++ is not good for POSIX.

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:165]: It's arguable to some extent. One may argue that Backbone Router without Multicast routing is also useful, thus should keep running. 
But agree that VerifyOrDie should be better for disclosing issues. Will change.

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:130]: Is it necessary for POSIX?

File: src/posix/platform/platform-posix.h
Comment (simonlingoogle) [reply src/posix/platform/platform-posix.h]: I will fix to use WARN.

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:83]: Now that we use VerifyOrDie for `InitMulticastRouterSock`, Enable never fail.

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp:130]: changed to use OT_MAX

File: src/posix/platform/backbone.cpp
Comment (bukepo) [inline src/posix/platform/backbone.cpp:65]: As discussed, It's possible to simplify the design according to a good kernel version, and I think we should update the current implementation which is designed according to work with a buggy kernel.

Comment (simonlingoogle) [reply src/posix/platform/backbone.cpp:65]: @jwhui @librasungirl @bukepo 

According to our investigation, latest 5.x Linux kernel supports (*, G) multicasting, meantime 4.x Linux and macOS(BSD) does not support (*, G) multicasting, but only support (S, G) multicasting. 

What MLR feature needs for the PBBR is (*, G) multicast routing, thus we need to implement (*, G) multicast routing using (S, G) multicast routing in the POSIX platform code.

Even though it's possible to use a simpler implementation using (*, G) multicast routing provided by latest Linux kernel, I think **the current implementation is valuable because it's more portable**. It works on a wide range of Linux kernel versions and can be ported to support macOS.

Comment (simonlingoogle) [reply src/posix/platform/backbone.cpp:65]: > I think we should update the current implementation which is designed according to work with a buggy kernel.

I am not sure if it's fair to say that the current implementation is designed to work with a buggy kernel.

The current implementation uses a set of APIs that are provided by many POSIX platforms, including Linux, macOS, BSD. 
It does not use [MRT6_ADD_MFC_PROXY](
https://github.com/torvalds/linux/blob/3d5e28bff7ad55aea081c1af516cc1c94a5eca7d/include/uapi/linux/mroute6.h#L32) which is only provided by latest Linux to add a (\*, G) MFC entry, which is not well documented and not proven to work (I had a version using MRT6_ADD_MFC_PROXY but it didn't work as intended).

That being said, I am not against a future enhancement that utilizes MRT6_ADD_MFC_PROXY for Linux kernel versions that has working  (\*, G) MFC. But we should keep this implementation for portability. 

Thoughts?

Comment (librasungirl) [reply src/posix/platform/backbone.cpp:65]: It is a good suggestion to do investigation on (*,G) which has some obvious advantages.
I may suggest do it later as future optimization without blocking this PR review.

Thoughts?
@bukepo

**Review by bukepo — APPROVED** [review]

File: src/posix/platform/multicast_routing.hpp
Comment (jwhui) [inline src/posix/platform/multicast_routing.hpp:40]: Are these used?

File: src/posix/platform/multicast_routing.cpp
Comment (jwhui) [inline src/posix/platform/multicast_routing.cpp:219]: Is it possible to use sizeof something here?

File: src/posix/platform/multicast_routing.hpp
Comment (jwhui) [inline src/posix/platform/multicast_routing.hpp:174]: nit: colocate with other methods above?

File: src/posix/platform/multicast_routing.cpp
Comment (jwhui) [inline src/posix/platform/multicast_routing.cpp:434]: Declare as `const`?

File: src/posix/platform/multicast_routing.cpp
Comment (jwhui) [inline src/posix/platform/multicast_routing.cpp:126]: Declare as `const`?

File: src/posix/platform/multicast_routing.cpp
Comment (simonlingoogle) [reply src/posix/platform/multicast_routing.cpp]: Changed to use `sizeof(struct mrt6msg)`.

**Review by jwhui — APPROVED** [review]

**Pull request merged**
